src/pybind/mgr/orchestrator/module.py

```diff
@@ -456,24 +457,84 @@ def ukn(s):
         'Create OSD daemon(s) using a drive group spec')
-    def _apply_osd(self, all_available_devices=False, inbuf=None):
-        # type: (bool, Optional[str]) -> HandleCommandResult
+    def _apply_osd(self,
+                   all_available_devices: bool = False,
+                   preview: bool = False,
+                   service_id: Optional[str] = None,
+                   unmanaged: Optional[bool] = None,
+                   format: Optional[str] = 'plain',
+                   inbuf: Optional[str] = None) -> HandleCommandResult:
         """Apply DriveGroupSpecs to create OSDs"""
         usage = """
 Usage:
   ceph orch apply osd -i <json_file/yaml_file>
   ceph orch apply osd --use-all-devices
+  ceph orch apply osd --service-id <service_id> --preview
+  ceph orch apply osd --service-id <service_id> --unmanaged=True|False
 """
+
+        def print_preview(prev, format):
+            if format != 'plain':
+                return to_format(prev, format)
+            else:
+                table = PrettyTable(
+                    ['NAME', 'HOST', 'DATA', 'DB', 'WAL'],
+                    border=False)
+                table.align = 'l'
+                table.left_padding_width = 0
+                table.right_padding_width = 1
+                for data in prev:
+                    dg_name = data.get('drivegroup')
+                    hostname = data.get('host')
+                    for osd in data.get('data', {}).get('osds', []):
+                        db_path = '-'
+                        wal_path = '-'
+                        block_db = osd.get('block.db', {}).get('path')
+                        block_wal = osd.get('block.wal', {}).get('path')
+                        block_data = osd.get('data', {}).get('path', '')
+                        if not block_data:
+                            continue
+                        if block_db:
+                            db_path = data.get('data', {}).get('vg', {}).get('devices', [])
+                        if block_wal:
+                            wal_path = data.get('data', {}).get('wal_vg', {}).get('devices', [])
+                        table.add_row((dg_name, hostname, block_data, db_path, wal_path))
+                out = table.get_string()
+                if not out:
+                    out = "No pending deployments."
+                return out
+
+        if inbuf or all_available_devices and service_id:
+            return HandleCommandResult(-errno.EINVAL, stderr=usage)
+
+        if preview and not service_id:
+            # get all stored drivegroups and print
+            prev = self.preview_drivegroups()
+            return HandleCommandResult(stdout=print_preview(prev, format))
+
+        if service_id and preview:
+            # get specified drivegroup and print
+            prev = self.preview_drivegroups(service_id)
+            return HandleCommandResult(stdout=print_preview(prev, format))
```

Comment:
this totally breaks

```
ceph orch apply osd --use-all-devices --preview
```
and
```
ceph orch apply osd -i <json_file/yaml_file> --preview
```
Can we instead move the `preview` flag into `apply_drivegroups` like so:

```
... = apply_drivegroups(dg_specs, preview)
```
?

Reply:
Mh, right.. This would need to implicitly set `unmanaged->True`..

Putting the preview flag in `apply_drivegroups` is not the right approach though.. I'll come up with something.

Reply:
all available options work now in conjunction with the --preview flag..

When `all-available-devices or inbuf` is used, we implicitly set the unmanaged flag to true.

I kinda dislike the way of doing it, but I don't see another option that covers all the required usecase right now.